csp/impl/perspective_common.py

```diff
@@ -0,0 +1,130 @@
+    return _PERSPECTIVE_3
+
+
+def perspective_type_map():
+    """Return the mapping of standard python types to perspective types"""
+    return {
+        str: "string",
+        float: "float",
+        int: "integer",
+        date: "date",
+        datetime: "datetime",
+        bool: "boolean",
+    }
+
+
+def datetime_to_perspective(dt: datetime) -> int:
+    """Convert a python datetime to an integer number of milliseconds for perspective"""
+    return int(dt.timestamp() * 1000)
+
+
+def date_to_perspective(d: date) -> int:
+    """Convert a python date to an integer number of milliseconds for perspective"""
+    return int(datetime(year=d.year, month=d.month, day=d.day, tzinfo=pytz.UTC).timestamp() * 1000)
+
+
+@csp.node
+def apply_updates(
+    table: perspective.Table,
+    data: {object: csp.ts[object]},
+    index_col: str,
+    time_col: str,
+    throttle: timedelta,
+    localize: bool,
+    static_records: dict = None,
+):
+    with csp.alarms():
+        alarm = csp.alarm(bool)
+
+    with csp.state():
+        s_buffer = []
+        s_has_time_col = False
+        s_datetime_cols = set()
+        s_date_cols = set()
+
+    with csp.start():
+        if throttle > timedelta(0):
+            csp.schedule_alarm(alarm, throttle, True)
+        s_has_time_col = time_col and time_col not in data.keys()
+        if _PERSPECTIVE_3:
+            s_datetime_cols = set([c for c, t in table.schema().items() if t == "datetime"])
+            s_date_cols = set([c for c, t in table.schema().items() if t == "date"])
+        else:
+            s_datetime_cols = set([c for c, t in table.schema().items() if t == datetime])
+
+    with csp.stop():
+        if len(s_buffer) > 0:
+            table.update(s_buffer)
+
+    if csp.ticked(data):
+        new_rows = {}
+        for (idx, col), value in data.tickeditems():
+            if idx not in new_rows:
+                row = {}
+                new_rows[idx] = row
+                if index_col:
+                    row[index_col] = idx
+                if s_has_time_col:
+                    if localize or _PERSPECTIVE_3:
+                        row[time_col] = pytz.utc.localize(csp.now())
+                    else:
+                        row[time_col] = csp.now()
+                    if _PERSPECTIVE_3:
+                        row[time_col] = datetime_to_perspective(row[time_col])
+            else:
+                row = new_rows[idx]
+
+            if (localize or _PERSPECTIVE_3) and col in s_datetime_cols and value.tzinfo is None:
+                row[col] = pytz.utc.localize(value)
+            else:
+                row[col] = value
+
+            if _PERSPECTIVE_3:
+                if col in s_datetime_cols:
+                    row[col] = datetime_to_perspective(row[col])
+                if col in s_date_cols:
+                    row[col] = date_to_perspective(row[col])
+
+        if static_records:
+            for idx, row in new_rows.items():
```

Comment:
not clear to me what static records is a dict of.  I would think its default values per column, but looks like its looking up by key here

Reply:
Its a dict of values that don't change, keyed by the "index" of the pandas dataframe (in the case where the data frame contains both csp edges and non-changing values of various types). For every tick of the csp edges, we want to propagate the "static values" corresponding to the same index.